pkg/csi/cinder/controllerserver.go

```diff
@@ -110,31 +112,56 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	}
 	content := req.GetVolumeContentSource()
 	var snapshotID string
-	var sourcevolID string
+	var sourceVolID string
+	var sourceBackupID string
 
 	if content != nil && content.GetSnapshot() != nil {
 		snapshotID = content.GetSnapshot().GetSnapshotId()
-		_, err := cloud.GetSnapshotByID(snapshotID)
-		if err != nil {
-			if cpoerrors.IsNotFound(err) {
-				return nil, status.Errorf(codes.NotFound, "VolumeContentSource Snapshot %s not found", snapshotID)
+
+		snap, snapErr := cs.Cloud.GetSnapshotByID(snapshotID)
+		if snapErr != nil && !cpoerrors.IsNotFound(snapErr) {
+			return nil, snapErr
+		}
+		// If the snapshot exists but is not yet available, fail.
+		if snapErr == nil && snap.Status != "available" {
+			return nil, status.Errorf(codes.NotFound, "VolumeContentSource Snapshot %s is not yet available. status: %s", snapshotID, snap.Status)
+		}
+
+		// In case a snapshot is not found
+		// check if a Backup with the same ID exists
+		if cpoerrors.IsNotFound(snapErr) {
+			back, err := cs.Cloud.GetBackupByID(snapshotID)
+			if err != nil {
+				//If there is an error getting the backup as well, fail.
+				return nil, status.Errorf(codes.NotFound, "VolumeContentSource Snapshot or Backup with ID %s not found", snapshotID)
+			}
+			if back.Status != "available" {
+				// If the backup exists but is not yet available, fail.
+				return nil, status.Errorf(codes.NotFound, "VolumeContentSource Backup %s is not yet available. status: %s", snapshotID, back.Status)
 			}
-			return nil, status.Errorf(codes.Internal, "Failed to retrieve the snapshot %s: %v", snapshotID, err)
+			// If an available backup is found, create the volume from the backup
+			sourceBackupID = snapshotID
+			snapshotID = ""
 		}
+
 	}
 
 	if content != nil && content.GetVolume() != nil {
-		sourcevolID = content.GetVolume().GetVolumeId()
-		_, err := cloud.GetVolume(sourcevolID)
+		sourceVolID = content.GetVolume().GetVolumeId()
+		_, err := cloud.GetVolume(sourceVolID)
 		if err != nil {
 			if cpoerrors.IsNotFound(err) {
-				return nil, status.Errorf(codes.NotFound, "Source Volume %s not found", sourcevolID)
+				return nil, status.Errorf(codes.NotFound, "Source Volume %s not found", sourceVolID)
 			}
-			return nil, status.Errorf(codes.Internal, "Failed to retrieve the source volume %s: %v", sourcevolID, err)
+			return nil, status.Errorf(codes.Internal, "Failed to retrieve the source volume %s: %v", sourceVolID, err)
 		}
 	}
 
-	vol, err := cloud.CreateVolume(volName, volSizeGB, volType, volAvailability, snapshotID, sourcevolID, &properties)
+	vol, err := cloud.CreateVolume(volName, volSizeGB, volType, volAvailability, snapshotID, sourceVolID, sourceBackupID, &properties)
+	// When creating a volume from a backup, the response does not include the backupID.
+	if sourceBackupID != "" {
+		vol.BackupID = &sourceBackupID
+	}
```

Comment:
It can't return a backupID since the returned object comes from the actual cinder volume object, which doesn't have a field to store what backup it was created from :sweat_smile:

Reply:
All clear, thanks!